### ISSUE TITLE
TableBatchOperation should be Thread Safe

### ISSUE THREAD

**Issue by kzehrer** [Azure/azure-sdk-for-net#1096]
In order to be able to utilize the TPL and/or PLINQ to efficiently setup large batch operations, the TableBatchOperation class should be thread safe for adding operations to the batch.

**Comment by veena-udayabhanu** [comment]
Can you give us more information on your scenario? A batch operation can contain at most 100 operations. Adding them in parallel will probably not help you much because adding an operation is a very small operation.

**Comment by kzehrer** [comment]
Sorry, this email got buried and I did not see it until now.  While 100 operations may not be a very large number, which it’s not, in a high-performance environment such as a cloud site, even 100 operations being set in a batch can benefit from using Parallel.ForEach as opposed to a serial foreach loop.

 

Currently, using a Parallel.ForEach to setup the operations in a TableBatchOperation will lead to non-deterministic results because the class is not thread-safe.

 

So, if I have 400 operations I want to setup, and I am going to break those down into 4 batches with 100 operations each, I can set up 4 batches and fill those in parallel, but 4 is not a number you would run parallel process on – more overhead to setup than benefit.  So because you wouldn’t parallel process 4 operations, and you CAN’T parallel process setting the operations in the batch, you are now running 400 sequential batch operation setups.  With that same thought in mind, if you were to choose a number of around 16 (or even 32) as being the point where you consider the number of operations benefitting from setting up a parallel processing operations, the example above could grow to saying that you won’t really be able to get out of serial/sequential processing until the number of operations gets over 1600 (or 3200).  Now the number is starting to get pretty big and inefficient.  

 

 

Thanks!

**Comment by shahabhijeet** [comment]
@kzehrer please open a new issue if the problem still exists/repros.
